# Data Cleaning

## Import packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re
%matplotlib inline

Read in csv and examine data

In [2]:
combine = pd.read_csv('../data/nfl_combine_data.csv')

In [3]:
# Look at descriptive stats
print(combine.shape, '\n')
print(combine.info())
combine.describe().T

(9950, 16) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9950 entries, 0 to 9949
Data columns (total 16 columns):
combine_year            9950 non-null int64
first_name              9950 non-null object
last_name               9950 non-null object
college                 9950 non-null object
position                9950 non-null object
height_inches           9950 non-null float64
weight_lbs              9950 non-null int64
hand_size_inches        8400 non-null float64
arm_length_inches       8082 non-null float64
40_yard_dash            9073 non-null float64
bench_press_reps        6779 non-null float64
vertical_leap_inches    8050 non-null float64
broad_jump_inches       7903 non-null float64
3_cone_drill            4518 non-null float64
20_yard_shuttle         7176 non-null float64
60_yard_shuttle         3169 non-null float64
dtypes: float64(10), int64(2), object(4)
memory usage: 1.2+ MB
None


count         mean        std      min      25%  \
combine_year          9950.0  2002.204824   9.302148  1987.00  1994.00   
height_inches         9950.0    73.735640   2.645112    64.90    71.88   
weight_lbs            9950.0   240.282513  45.046373   142.00   203.00   
hand_size_inches      8400.0     9.528956   0.629254     7.13     9.13   
arm_length_inches     8082.0    32.221633   1.498680    25.63    31.25   
40_yard_dash          9073.0     4.830682   0.309814     4.21     4.59   
bench_press_reps      6779.0    19.833014   6.540305     1.00    15.00   
vertical_leap_inches  8050.0    32.001801   4.203431    17.50    29.00   
broad_jump_inches     7903.0   112.305580   9.306169     7.00   106.00   
3_cone_drill          4518.0     7.346076   0.446751     6.34     7.01   
20_yard_shuttle       7176.0     4.403852   0.268681     3.73     4.20   
60_yard_shuttle       3169.0    11.674550   0.425484    10.43    11.38   

                          50%      75%      max  
combine_year          2002.00  2011.00  2017.00  
height_inches           74.00    75.75    82.40  
weight_lbs             232.00   275.00   387.00  
hand_size_inches         9.50    10.00    11.88  
arm_length_inches       32.25    33.25    38.50  
40_yard_dash             4.76     5.05     6.12  
bench_press_reps        20.00    24.00    51.00  
vertical_leap_inches    32.00    35.00    46.00  
broad_jump_inches      113.00   119.00   147.00  
3_cone_drill             7.26     7.62     9.61  
20_yard_shuttle          4.37     4.57     5.68  
60_yard_shuttle         11.65    11.93    13.91

3_cone_drill, 60_yard_shuttle have many missing values

In [9]:
combine.head()

combine_year  first_name  last_name          college position  \
0          2017       jamal      adams  louisiana_state       db   
1          2017  montravius      adams           auburn       dl   
2          2017      rodney      adams    south_florida       wr   
3          2017      quincy  adeboyejo      mississippi       wr   
4          2017       brian      allen             utah       db   

   height_inches  weight_lbs  hand_size_inches  arm_length_inches  \
0          71.63         214              9.25              33.38   
1          75.63         304              9.25              32.75   
2          73.25         189              9.00              32.00   
3          74.75         197              9.38              31.75   
4          74.88         215             10.00              34.00   

   40_yard_dash  bench_press_reps  vertical_leap_inches  broad_jump_inches  \
0          4.56              18.0                  31.5              120.0   
1          4.87              22.0                  29.0              108.0   
2          4.44               8.0                  29.5              125.0   
3          4.42               8.0                  34.5              123.0   
4          4.48              15.0                  34.5              117.0   

   3_cone_drill  20_yard_shuttle  60_yard_shuttle       player_name  
0          6.96             4.13              NaN       jamal adams  
1          7.62              NaN              NaN  montravius adams  
2          6.98             4.28            11.39      rodney adams  
3          6.73             4.14              NaN  quincy adeboyejo  
4          6.64             4.34              NaN       brian allen

In [8]:
# Add full name as a column
combine['player_name'] = combine.first_name + ' ' + combine.last_name

In [6]:
# Check years of data
combine.combine_year.unique()

array([2017, 2016, 2015, 2014, 2013, 2012, 2011, 2010, 2009, 2008, 2007,
       2006, 2005, 2004, 2003, 2002, 2001, 2000, 1999, 1998, 1997, 1996,
       1995, 1994, 1993, 1992, 1991, 1990, 1989, 1988, 1987])

In [7]:
combine.groupby('combine_year').count()

first_name  last_name  college  position  height_inches  \
combine_year                                                            
1987                 267        267      267       267            267   
1988                 322        322      322       322            322   
1989                 308        308      308       308            308   
1990                 324        324      324       324            324   
1991                 428        428      428       428            428   
1992                 438        438      438       438            438   
1993                 338        338      338       338            338   
1994                 315        315      315       315            315   
1995                 310        310      310       310            310   
1996                 296        296      296       296            296   
1997                 319        319      319       319            319   
1998                 303        303      303       303            303   
1999                 377        377      377       377            377   
2000                 258        258      258       258            258   
2001                 266        266      266       266            266   
2002                 270        270      270       270            270   
2003                 259        259      259       259            259   
2004                 274        274      274       274            274   
2005                 285        285      285       285            285   
2006                 279        279      279       279            279   
2007                 271        271      271       271            271   
2008                 267        267      267       267            267   
2009                 285        285      285       285            285   
2010                 287        287      287       287            287   
2011                 302        302      302       302            302   
2012                 326        326      326       326            326   
2013                 293        293      293       293            293   
2014                 304        304      304       304            304   
2015                 741        741      741       741            741   
2016                 319        319      319       319            319   
2017                 319        319      319       319            319   

              weight_lbs  hand_size_inches  arm_length_inches  40_yard_dash  \
combine_year                                                                  
1987                 267               237                237           237   
1988                 322               316                  1           272   
1989                 308                 0                  0           270   
1990                 324               314                314           287   
1991                 428               426                426           371   
1992                 438               421                421           364   
1993                 338               337                337           270   
1994                 315               315                315           249   
1995                 310               309                309           218   
1996                 296               296                296           222   
1997                 319               319                319           245   
1998                 303               302                302           225   
1999                 377               274                274           377   
2000                 258               231                231           258   
2001                 266               237                237           266   
2002                 270               242                242           270   
2003                 259               238                238           259   
2004                 274               248                248           274   
2005                 285

In [10]:
combine.loc[combine.combine_year == 2015].isnull().sum()

combine_year              0
first_name                0
last_name                 0
college                   0
position                  0
height_inches             0
weight_lbs                0
hand_size_inches        428
arm_length_inches       429
40_yard_dash              5
bench_press_reps        517
vertical_leap_inches    485
broad_jump_inches       486
3_cone_drill            516
20_yard_shuttle         508
60_yard_shuttle         648
player_name               0
dtype: int64

## Addressing 2015 data

- 2015 has 741 unique attendees. There are only 320, so we need to get rid of the extra attendees in our dataset.
- Load in actual attendee data
- Split first and last names into 2 columns
- Remove special characters
- Make everything lower case
- Inner join original combine and attendee data
- Drop 2015 rows from original combine data
- Concat joined result with combine data

In [11]:
combine = pd.read_csv('../data/nfl_combine_data.csv')
combine['player_name'] = combine.first_name + ' ' + combine.last_name

In [12]:
# Read in 2015 data with 320 attendees
df15 = pd.read_excel('../data/2015_combine_data.xlsx')
# drop unnecessary rows
cols = ['Player','Year','Pos','School']
df15 = df15[cols]
df15.rename(columns ={'Player':'player_name','Year':'combine_year', 'Pos':'position', 'School': 'college'},inplace=True)

In [13]:
combine15 = combine.loc[combine.combine_year == 2015]

In [14]:
df15.head()

player_name  combine_year position        college
0  Ameer Abdullah          2015       RB       Nebraska
1  Nelson Agholor          2015       WR            USC
2       Jay Ajayi          2015       RB      Boise St.
3  Kwon Alexander          2015      OLB            LSU
4    Mario Alford          2015       WR  West Virginia

In [15]:
combine15.head()

combine_year first_name  last_name              college position  \
638          2015      ameer   abdullah             nebraska       rb   
639          2015     nelson    agholor  southern_california       wr   
640          2015    malcolm      agnew    southern_illinois       rb   
641          2015        jay      ajayi          boise_state       rb   
642          2015    brandon  alexander      central_florida       db   

     height_inches  weight_lbs  hand_size_inches  arm_length_inches  \
638          68.75         205              8.63              30.00   
639          72.13         198              9.25              32.25   
640          70.00         202               NaN                NaN   
641          71.75         221             10.00              32.00   
642          74.00         195               NaN                NaN   

     40_yard_dash  bench_press_reps  vertical_leap_inches  broad_jump_inches  \
638          4.60              24.0                  42.5              130.0   
639          4.42              12.0                   NaN                NaN   
640          4.59               NaN                   NaN                NaN   
641          4.57              19.0                  39.0              121.0   
642          4.59               NaN                   NaN                NaN   

     3_cone_drill  20_yard_shuttle  60_yard_shuttle        player_name  
638          6.79             3.95            11.18     ameer abdullah  
639           NaN              NaN              NaN     nelson agholor  
640           NaN              NaN              NaN      malcolm agnew  
641          7.10             4.10            11.10          jay ajayi  
642           NaN              NaN              NaN  brandon alexander

## Cleaning Functions

In [16]:
regex = re.compile("[@_!#$%^&*()<>?/\|}{~:`'']") 

In [17]:
# Remove special characters
def remove_special(df):
    df = df.applymap(lambda x: re.sub("[@_!#$%^&*.<>?/\|}{~:`'']", '', x) if type(x) == str else x)

def lowercase(df, col):
    df[col] = df[col].apply(lambda x: x.lower())
#         combine15.player = combine15.player_name.apply(lambda x: x.lower())

def clean_name(df, col):
    df[['first_name','last_name']] = df[col].str.split(" ",n=1,expand=True)

# Clean combine_year record that lists '5' instead of '2005'
def replace_5(df):
    df.loc[df.combine_year == 5, 'combine_year'] = 2005

# Add 'events_missed' columns
def events_missed():
    pass
# Add 'conference' column

# Position standardization

# Regression imputation for missing values

In [18]:
replace_5(combine15)
remove_special(combine15) # doesn't work
lowercase(combine15,'player_name')
clean_name(combine15,'player_name')

/Users/flatironschooldc2/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/indexing.py:494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
/Users/flatironschooldc2/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/flatironschooldc2/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/frame.py:3494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice f

In [19]:
remove_special(df15) # doesn't work
lowercase(df15, 'player_name')
clean_name(df15,'player_name')

In [20]:
df15 = df15.applymap(lambda x: re.sub("[@_!#$%^&*.<>?/\|}{~:`'']", '', x) if type(x) == str else x)
combine15 = combine15.applymap(lambda x: re.sub("[@_!#$%^&*.<>?/\|}{~:`'']", '', x) if type(x) == str else x)

In [21]:
print(df15[~df15.player_name.isin(combine15.player_name)].player_name.unique())
len(df15[~df15.player_name.isin(combine15.player_name)].player_name.unique())

['trenton brown' 'michael burton' 'yannick cudjoe-virgil'
 'durell eskridge' 'mario edwards, jr' 'donatella luckett'
 'zack wagenmann']


7

In [22]:
merged_df = pd.merge(df15, combine15, how='left',copy=False, on = ['first_name', 'last_name', 'combine_year', 'position'])

In [23]:
merged_df = merged_df.drop(['player_name_y','college_y'],axis=1).rename(columns={'player_name_x':'player_name','college_x':'college'})

In [24]:
# create first and last name columns for OG combine data
clean_name(combine, 'player_name')

In [25]:
# cols = combine
merged_df = merged_df[['combine_year', 'player_name','first_name', 'last_name', 'college', 'position', 'height_inches',
       'weight_lbs', 'hand_size_inches', 'arm_length_inches', '40_yard_dash',
       'bench_press_reps', 'vertical_leap_inches', 'broad_jump_inches',
       '3_cone_drill', '20_yard_shuttle', '60_yard_shuttle']]

combine = combine[['combine_year', 'player_name','first_name', 'last_name', 'college', 'position', 'height_inches',
       'weight_lbs', 'hand_size_inches', 'arm_length_inches', '40_yard_dash',
       'bench_press_reps', 'vertical_leap_inches', 'broad_jump_inches',
       '3_cone_drill', '20_yard_shuttle', '60_yard_shuttle']]

In [26]:
# drop 2015 rows from combine df
index_drop_list = combine.loc[combine.combine_year == 2015].index

In [27]:
combine.head()

combine_year       player_name  first_name  last_name          college  \
0          2017       jamal adams       jamal      adams  louisiana_state   
1          2017  montravius adams  montravius      adams           auburn   
2          2017      rodney adams      rodney      adams    south_florida   
3          2017  quincy adeboyejo      quincy  adeboyejo      mississippi   
4          2017       brian allen       brian      allen             utah   

  position  height_inches  weight_lbs  hand_size_inches  arm_length_inches  \
0       db          71.63         214              9.25              33.38   
1       dl          75.63         304              9.25              32.75   
2       wr          73.25         189              9.00              32.00   
3       wr          74.75         197              9.38              31.75   
4       db          74.88         215             10.00              34.00   

   40_yard_dash  bench_press_reps  vertical_leap_inches  broad_jump_inches  \
0          4.56              18.0                  31.5              120.0   
1          4.87              22.0                  29.0              108.0   
2          4.44               8.0                  29.5              125.0   
3          4.42               8.0                  34.5              123.0   
4          4.48              15.0                  34.5              117.0   

   3_cone_drill  20_yard_shuttle  60_yard_shuttle  
0          6.96             4.13              NaN  
1          7.62              NaN              NaN  
2          6.98             4.28            11.39  
3          6.73             4.14              NaN  
4          6.64             4.34              NaN

In [28]:
combine = combine.drop(index_drop_list)

In [29]:
# Union datasets together
combine = pd.concat([combine,merged_df])

In [30]:
combine.reset_index(inplace=True)
combine.shape

(9531, 18)

## Adding Events Missed Information

In [33]:
combine.isnull().sum()

index                      0
combine_year               0
player_name                0
first_name                 0
last_name                  0
college                    0
position                   0
height_inches            322
weight_lbs               322
hand_size_inches        1444
arm_length_inches       1761
40_yard_dash            1194
bench_press_reps        2976
vertical_leap_inches    1737
broad_jump_inches       1883
3_cone_drill            5238
20_yard_shuttle         2588
60_yard_shuttle         6455
dtype: int64

In [32]:
combine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9531 entries, 0 to 9530
Data columns (total 18 columns):
index                   9531 non-null int64
combine_year            9531 non-null int64
player_name             9531 non-null object
first_name              9531 non-null object
last_name               9531 non-null object
college                 9531 non-null object
position                9531 non-null object
height_inches           9209 non-null float64
weight_lbs              9209 non-null float64
hand_size_inches        8087 non-null float64
arm_length_inches       7770 non-null float64
40_yard_dash            8337 non-null float64
bench_press_reps        6555 non-null float64
vertical_leap_inches    7794 non-null float64
broad_jump_inches       7648 non-null float64
3_cone_drill            4293 non-null float64
20_yard_shuttle         6943 non-null float64
60_yard_shuttle         3076 non-null float64
dtypes: float64(11), int64(2), object(5)
memory usage: 1.3+ MB


## Regression Imputation